lib/data/repository/default_marvell_repository.dart

```diff
@@ -0,0 +1,113 @@
+import 'package:dio/dio.dart';
+import 'package:marvelapp_flutter/domain/entities/character.dart';
+import 'package:marvelapp_flutter/domain/error_handling/exceptions.dart';
+import 'package:marvelapp_flutter/domain/repositories/marvell_repository.dart';
+import 'package:marvelapp_flutter/domain/entities/series.dart';
+import 'package:marvelapp_flutter/data/data_sources/local/character_data_source.dart';
+import 'package:marvelapp_flutter/data/data_sources/remote/remote_data_source.dart';
+
+class DefaultMarvellRepository extends MarvellRepository {
+  late CharacterDataSource localDataSource;
+  late RemoteDataSource remoteDataSource;
+
+  static final DefaultMarvellRepository _repository = DefaultMarvellRepository._internal();
+
+  DefaultMarvellRepository._internal();
+
+  factory DefaultMarvellRepository(CharacterDataSource localDataSource, RemoteDataSource remoteDataSource) {
+    _repository.localDataSource = localDataSource;
+    _repository.remoteDataSource = remoteDataSource;
+    return _repository;
+  }
+
+  @override
+  Future<Character> getCharacterDetail(String characterId) async {
+    try {
+      var httpResponse = await remoteDataSource.getCharacterDetail(characterId);
+      if (httpResponse.response.statusCode == 200) {
+        var data = httpResponse.data.data;
+        Character? tempCharacter = data?.results?.map((item) => item.fromApiToCharacter("standard_xlarge")).single;
+        Character character = tempCharacter as Character;
+        return character;
+      } else {
+        throw DataRetrieveException();
+      }
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  @override
+  Future<List<Character>> getCharacters([int offset = 0]) async {
+    if (offset == 0) {
+      List<Character> characters = List.empty();
```

Comment:
```suggestion
      List<Character> characters;
```

Reply:
only in this case it work. because in try and catch blocks we assign to characters something.

but in this case class Repository {

  List<Character> getCharacters() {
    List<Character> list;

    return list;
  }
it dont work. because list is null.
i can made working code snippet above like this

class Repository {

  List<Character>? getCharacters() {
    List<Character>? list;

    return list;
  }

в дарте не дают работать с неинициализированными переменными. тут либо late, либо переменная должна быть нулабельной. с динамическим типом компилируется, но потом можно влететь в ошибку что переменная не инициализирована, на уровне поля класса динамическая переменная становится равна null  как в java,  но я так понимаю что лучше это меньше использовать.

done.